gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+                state.request = None;
+                return Ok(());
+            }
+
+            VIRTIO_GPIO_DIRECTION_IN => config.set_direction_offset(Direction::Input, gpio as u32),
+            VIRTIO_GPIO_DIRECTION_OUT => {
+                config.set_direction(Direction::Output);
+                config.set_output_values(&[gpio as u32], &[value as i32]);
+            }
+
+            _ => return Err(Error::GpioDirectionInvalid(value)),
+        };
+
+        if let Some(request) = &state.request {
+            request
+                .reconfigure_lines(&config)
+                .map_err(Error::GpiodFailed)?;
+        } else {
+            let rconfig = RequestConfig::new().map_err(Error::GpiodFailed)?;
+
+            rconfig.set_consumer("vhu-gpio");
+            rconfig.set_offsets(&[gpio as u32]);
+
+            state.request = Some(Arc::new(
+                self.chip
+                    .request_lines(&rconfig, &config)
+                    .map_err(Error::GpiodFailed)?,
+            ));
+        }
+
+        Ok(())
+    }
+
+    fn get_value(&self, gpio: u16) -> Result<u8> {
+        let state = &self.state[gpio as usize].read().unwrap();
+
+        if let Some(request) = &state.request {
+            Ok(request.get_value(gpio as u32).map_err(Error::GpiodFailed)? as u8)
+        } else {
+            Err(Error::GpioDirectionInvalid(
+                VIRTIO_GPIO_DIRECTION_NONE as u32,
+            ))
+        }
+    }
+
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()> {
+        let state = &self.state[gpio as usize].read().unwrap();
+
+        // Direction change can follow value change, don't fail here for invalid
+        // direction.
+        if let Some(request) = &state.request {
+            request
+                .set_value(gpio as u32, value as i32)
+                .map_err(Error::GpiodFailed)?;
+        }
+
+        Ok(())
+    }
+
+    fn set_irq_type(&self, gpio: u16, value: u16) -> Result<()> {
+        let state = &mut self.state[gpio as usize].write().unwrap();
+        let mut config = LineConfig::new().map_err(Error::GpiodFailed)?;
+
+        match value as u16 {
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_RISING => config.set_edge_detection(Edge::Rising),
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_FALLING => config.set_edge_detection(Edge::Falling),
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH => config.set_edge_detection(Edge::Both),
+
+            // Drop the buffer.
+            VIRTIO_GPIO_IRQ_TYPE_NONE => {
+                state.buffer = None;
+                return Ok(());
+            }
+
+            // Only edge IRQs are supported for now.
+            _ => return Err(Error::GpioIrqTypeNotSupported(value)),
+        };
+
+        // Allocate the buffer and configure the line for interrupt.
+        if state.request.is_none() {
+            Err(Error::GpioLineRequestNotConfigured)
+        } else {
+            state.buffer = Some(EdgeEventBuffer::new(1).map_err(Error::GpiodFailed)?);
```

Comment:
I would make the size of the event buffer a paramater of this function.  Setting the buffer size to one regardless of the underlying device the line is attached to seems quite stiff to me.

Reply:
Which function? EdgeEventBuffer::new() ? or set_irq_type() ?

The argument X to EdgeEventBuffer::new(X) here is the capacity for the number of events we want to get for the line. Since these commands are sent by guest's in the regular case, it should always be one. Since the kernel never takes a list of interrupts for a line, but just a single interrupt.

Reply:
I was referring to set_irq_type().  A guest might want to receive more than one IRQ per call of wait_for_interrupt() or a device attached to a specific line may generate several interrupts per transfer.  But that could be handled in the guest driver as well and how this kind of situation should be handled is something the spec may be clarifying. 

Regardless, I think a good dose of in-line documentation is needed here.

Reply:
The specification doesn't handle with multiple interrupts as one. It is designed in a similar way to any hardware IP, which gives only one interrupt per line at once.

Sure I will add a comment on why "1" is used here.